### PR TITLE
Add Optimized GPT

### PR DESCRIPTION
**Pull request by nik-mosaic**
Adds code for an optimized version of MosaicGPT. The user can specify cfg.gpt_block = 'optimized' to run this code, and cfg.gpt_block = 'standard' (or omit it) to run the standard code.

This PR contains:
  - In `src/mosaic_gpt.py` and `src/models/layers/gpt_blocks.py`: Modeling changes to enable the optimized GPT: moving first layernorm of GPT blocks 2 - n to the end of blocks 1 - n-1, and replacing the final layernorm with `ln_f` with `ln_i`. These should be math-equivalent, so if you choose not to use cfg.gpt_block = 'optimized', your model will be unchanged.
  - In `src/models/layers/gpt_blocks.py`: The new OptimizedGPTBlock itself.
  - In `src/mosaic_gpt.py`: An option to use Fused Cross Entropy, which is enabled by default, installed via the standard `pip install .[llm]`. Otherwise, you can set `cfg.loss_fn=torch_crossentropy` to disable this and use standard torch.nn.CrossEntropy().
  - In the `csrc/` folder: C++/CUDA code for the one custom fusion we include. Since the HazyResearch DropoutAddLayerNorm does not support 30B and 70B models, we add code to support these models and require the user install it.
- In `llm/README.md` and `llm/csrc/README.md`: Installation instructions for dependencies for optimized MosaicGPT.

Along for the ride:
- Changes all vocab_sizes in YAMLs to 50304 (`icl_evals/yamls`, `yamls/mosaic_gpt/`, `mcloud/`)

Future Work:
- FusedMLP has been removed since it doesn't help performance on bf16 for CUDA < 11.8. In the future, when we support PyTorch 2.0 + CUDA 11.8, I will retest this.
- After this PR, I will make a pull request in the HazyResearch repo to add DropoutAddLayerNorm support for 30B and 70B model sizes . Then, we can remove the `csrc/` folder and include DropoutAddLayerNorm as a single line in a requirements file.
- Bake in DropoutAddLayerNorm into our docker images, like we've done with FlashAttention, so that it does not need to build every time (~20 min. build time).

### CODE REVIEW

File: examples/llm/src/models/ops/fused_dense.py
Comment (dakinggg) [inline examples/llm/src/models/ops/fused_dense.py:4]: remove?

File: examples/llm/src/models/ops/layer_norm.py
Comment (dakinggg) [inline examples/llm/src/models/ops/layer_norm.py:12]:
```suggestion
    DROPOUT_LAYER_NORM_INSTALLED = True
```
to be consistent

File: examples/llm/src/models/ops/fused_dense.py
Comment (dakinggg) [inline examples/llm/src/models/ops/fused_dense.py:18]: is this var used?

File: examples/llm/src/models/ops/layer_norm.py
Comment (dakinggg) [reply examples/llm/src/models/ops/layer_norm.py:12]: also, is this var used?

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [inline examples/llm/src/models/mosaic_gpt.py:92]:
```suggestion
        config_block = cfg.get('gpt_block', 'standard')
        if config_block == 'optimized':
            block_cls = gpt_blocks.OptimizedGPTBlock
        elif config_block == 'standard':
            block_cls = gpt_blocks.GPTBlock
        else:
            raise NotImplementedError('MosaicGPT `gpt_block` must be one of ['standard', 'optimized']
```

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dakinggg) [inline examples/llm/src/models/layers/gpt_blocks.py:36]: can you raise a nice error somewhere in the chain if this is going to fail on forward?

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dakinggg) [inline examples/llm/src/models/layers/gpt_blocks.py:87]: raise instead of assert

**Review by dakinggg — COMMENTED** [review]
This looks good to me as best I can tell. I did not review the "adapted/inspired" files. Let me know if there is anything specific in there you would like me to review. Would like to get another set of eyes for approval.

**Comment by dakinggg** [comment]
For testing, could you please show plots of
1) gpt-125 without this pr
2) gpt-125 standard with this pr
3) gpt-125 optimized with this pr

**Comment by dakinggg** [comment]
Sounds good, 1B model is fine too

File: examples/llm/README.md
Comment (dskhudia) [inline examples/llm/README.md:221]: installing the installing => installing the

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dskhudia) [reply examples/llm/src/models/layers/gpt_blocks.py:39]: It should fail on init and check_if_dense_gelu_dense_installed will raise the error.

File: examples/llm/src/models/mosaic_gpt.py
Comment (dskhudia) [inline examples/llm/src/models/mosaic_gpt.py:27]: It will fail here if not installed? This is not ideal.

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dskhudia) [inline examples/llm/src/models/layers/gpt_blocks.py:95]: Do these support `meta` device?

File: examples/llm/src/models/mosaic_gpt.py
Comment (dskhudia) [inline examples/llm/src/models/mosaic_gpt.py:290]: Is this supposed to be uncommented?

**Review by dskhudia — APPROVED** [review]

**Comment by dskhudia** [comment]
@nik-mosaic Also let us change vocab size for all yamls with this? currently I see `vocab_size: 50257` for 1b model.

File: examples/llm/README.md
Comment (dakinggg) [inline examples/llm/README.md:64]: can you take a todo to decide whether we should ship a docker image with all of this preinstalled?

File: examples/llm/yamls/mosaic_gpt/1b.yaml
Comment (abhi-mosaic) [inline examples/llm/yamls/mosaic_gpt/1b.yaml:26]: remove for now?

File: examples/llm/yamls/mosaic_gpt/1b.yaml
Comment (abhi-mosaic) [inline examples/llm/yamls/mosaic_gpt/1b.yaml:41]: also remove

File: examples/llm/src/models/mosaic_gpt.py
Comment (abhi-mosaic) [inline examples/llm/src/models/mosaic_gpt.py:23]:
```suggestion
from examples.llm.src.models.ops import CrossEntropyLoss as FusedCrossEntropyLoss
```

Could be a more descriptive name?

**Review by abhi-mosaic — COMMENTED** [review]
Few small edits but this looks great! Thank you for helping modularize the compments and the imports.

Last request before merging this -- since this is changing the `standard` architecture via the LN arrangement, can we redo a 1B run from last Friday, off this branch, and confirm that the convergence is still as expected? Should take <1 day on 4 nodes and I can help find allocation @nik-mosaic. Once that works, happy to approve and merge!

**Review by dakinggg — COMMENTED** [review]
The things from hazy research need to include the license (https://github.com/HazyResearch/flash-attention/blob/main/LICENSE)

The things from apex need to include the license (https://github.com/NVIDIA/apex/blob/master/LICENSE)

File: examples/llm/csrc/README.md
Comment (dakinggg) [reply examples/llm/csrc/README.md:18]: ^this suggestion did not really work cause of the backticks, but it is to add text that says "or omit model.gpt_block completely, which will default to standard"

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:285]: understood, I think as you have it is ok. Just wanted to make sure that change was intentional.

File: examples/llm/csrc/layer_norm/ln.h
Comment (vchiley) [inline examples/llm/csrc/layer_norm/ln.h:1]: [strongly consider adding all of these files as integrations (or convert the to a pip install)](https://mosaicml.slack.com/archives/C02KA7Y7NH0/p1678137737349809) instead of copy pasting thousands of lines of code which we will then need to maintain

**Review by vchiley — CHANGES_REQUESTED** [review]

File: examples/llm/src/models/ops/layer_norm.py
Comment (samhavens) [inline examples/llm/src/models/ops/layer_norm.py:33]: does this work? if we set this to `True` do we have RMSNorm? CC @bmosaicml

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (vchiley) [inline examples/llm/src/models/layers/gpt_blocks.py:87]: What does `prenorm=True` mean?
Does this mean make it a pre(layer)norm block?

Is GPTMLP and OptimizedGPTBlock exactly the same?
If yes, can we add testes to verify this? (along with bwd pass tests?)

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:72]: does `ln_i` mean initial ln or ln at index i?
this is a confusing name.

File: examples/llm/src/models/param_init_fns.py
Comment (vchiley) [inline examples/llm/src/models/param_init_fns.py:151]: why not combine this with the `elif isinstance(module, nn.LayerNorm): ` section?
its just a longer if statement.

**Pull request closed without merging**
